qlty-cloud/src/auth/auth_flow.rs

```diff
@@ -99,53 +76,91 @@ fn redirect_matches(expected: String, actual: String) -> bool {
     false
 }
 
-fn login_server(state: Arc<AppState>) -> Result<(Server, u16)> {
-    let listener = TcpListener::bind("127.0.0.1:0")?;
-    let port = listener.local_addr()?.port();
-    let server = HttpServer::new(move || {
-        App::new()
-            .app_data(Data::new(state.clone()))
-            .service(auth_flow)
-    })
-    .shutdown_timeout(0)
-    .workers(1)
-    .listen(listener)?
-    .run();
-    Ok((server, port))
-}
+fn run_handler(request: &Request, state: &AppState) -> Result<ResponseBox> {
+    let uri = request.url().parse::<Uri>()?;
+    let params: AuthFlowQueryParams;
+    if let Some(query) = uri.query() {
+        params = serde_querystring::from_str::<AuthFlowQueryParams>(query, ParseMode::UrlEncoded)
+            .with_context(|| "invalid parameter")?;
+    } else {
+        bail!("No query parameters found in request");
+    }
 
-pub fn launch_login_server(state: AppState) -> Result<u16> {
-    let handle = state.server_handle.clone();
-    let (server, port) = login_server(Arc::new(state))?;
+    if params.state != state.original_state {
+        bail!("State does not match original state");
+    }
 
-    handle.lock().unwrap().replace(server.handle());
+    if !redirect_matches(state.login_url.clone(), params.redirect_uri.clone()) {
+        bail!("Redirect URI does not match login URL");
+    }
 
-    thread::spawn(move || {
-        let _ = tokio::runtime::Builder::new_current_thread()
-            .enable_all()
-            .build()
-            .unwrap()
-            .block_on(server);
-    });
+    let token = Token::new(&state.credential_user);
```

Comment:
Calling it TokenStore would defeat the purpose of abstracting storage. At that point we could just dump the struct entirely and use regular methods to read/write from storage-- I had an initial version with this but removed it to avoid changing too much of the original implementation.

Ideally we would just have a simplified function oriented API:

```rust
// mod credential_store
fn read_token(user: &str) -> Result<String>
fn write_token(user: &str, value: &str) -> Result<()>
```

There's really no reason to create a token object / hydrate values since it's not used in any long-lived fashion. We throw away the Token object as soon as we read: https://github.com/qltysh/qlty/pull/1368/files#diff-62de191f887ff041d4cf140614b940a83d994dcde45fe5763f2063ebd076cf44R14-R18

Reply:
Ah yes, that makes sense. Agreed.